### PR TITLE
Rover steering wheel

### PR DESCRIPTION
**Pull request by noether**
This PR adds support to rovers with two actuators; namely, thrust (both directions) and steering wheel.

The four major (independent) points are:
- New firmware for the vehicle based on the existing firmware for rovers.
- We have added a NPS simulator for the rover. It behaves very similarly to a rover traveling at around 20km/h on a flat world.
- We have modified the "classic" GVF. Now you can call it from different firmware (fixed-wing and rover steering wheel).
- We have added UCM directory in flightplans and airframefiles for our configurations.

### CODE REVIEW

**Comment by noether** [comment]
Oh, I see that we forgot to implement the speed controller. We will upload it after a couple of tests. Then, we will move the current "Assisted mode" to "Nav mode". The assisted mode will disappear probably and we will have only two modes, manual and auto2/nav.

Anyway, there is a lot of code to check by you guys :P.

File: conf/messages.xml [hunk 1]
```diff
@@ -0,0 +1,3221 @@
+<?xml version="1.0"?>
+<!DOCTYPE protocol SYSTEM "messages.dtd">
```
Comment (gautierhattenberger) [inline conf/messages.xml:2]: remove this file from conf folder
if you need a new message, make a pull request in pprzlink and update the reference in paparazzi submodule

File: conf/modules/board_matek_f765_car.xml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+
+<module name="board_matek_f765_car" dir="boards">
```
Comment (gautierhattenberger) [inline conf/modules/board_matek_f765_car.xml:3]: what is the difference with "wing" version ?

File: conf/modules/fdm_rover.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+
+<module name="fdm_rover" dir="fdm">
+  <doc>
+    <description>
+
```
Comment (gautierhattenberger) [inline conf/modules/fdm_rover.xml:6]: please add a short description

File: conf/radios/UCM/T16SZ_SBUS_rover.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+-- average: (boolean) channel filtered through several frames (for discrete commands)
+-- min: minimum pulse length (micro-seconds)
+-- max: maximum pulse length (micro-seconds)
+-- neutral: neutral pulse length (micro-seconds)
+-- Note: a command may be reversed by exchanging min and max values
+-->
+
+<!DOCTYPE radio SYSTEM "../radio.dtd">
+<radio name="Futaba T16SZ with SBUS" data_min="1000" data_max="2000" sync_min ="5000" sync_max ="15000" pulse_type="POSITIVE">
+ <channel ctl="A" function="ROLL"     min="1000" neutral="1500" max="2000" average="0"/>
```
Comment (gautierhattenberger) [inline conf/radios/UCM/T16SZ_SBUS_rover.xml:48]: the `ctl` attribute is no longer needed and can be removed

File: conf/telemetry/steering_rover.xml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+<telemetry>
+
+
+  <process name="Main">
+
+    <mode name="default" key_press="d">
+      <message name="AUTOPILOT_VERSION"      period="11.1"/>
+      <message name="ROVER_STATUS"           period="1.2"/>
+      <message name="ROTORCRAFT_FP"          period="0.25"/>
+      <message name="STEERING_ROVER_DATA"    period="0.5"/>
```
Comment (gautierhattenberger) [inline conf/telemetry/steering_rover.xml:12]: maybe a new telemetry file is not needed and can be merged with default_rover.xml

File: conf/tools/pprzgcs.xml [hunk 6]
```diff
@@ -1,2 +0,0 @@
-<program command="$pprzgcs" name="PprzGCS" />
```
Comment (gautierhattenberger) [inline conf/tools/pprzgcs.xml:1]: please don't remove this file

File: sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c [hunk 7]
```diff
@@ -0,0 +1,102 @@
+/* rover_steering_guidance.c */
```
Comment (gautierhattenberger) [inline sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c:1]: please add paparazzi license header

File: sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  float delta = 0.0;
+  float speed = BoundSpeed(guidance_control.state_speed); 
+  float omega = guidance_control.gvf_omega; //GVF give us this omega
+
+  // ASSISTED guidance .....................................................
+  if (autopilot_get_mode() == AP_MODE_ASSISTED) {
+    if (fabs(omega)>0.0) {
+      delta = -atanf(omega * DRIVE_SHAFT_DISTANCE / speed);
+    }
+    delta *= 180/M_PI;
```
Comment (gautierhattenberger) [inline sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c:75]: if needed, their are `RadOfDeg` and `DegOfRad` macros

File: sw/airborne/firmwares/rover/main.c [hunk 9]
```diff
@@ -0,0 +1,3 @@
+// Empty... rovers only use main_ap...
```
Comment (gautierhattenberger) [inline sw/airborne/firmwares/rover/main.c:1]: why do you need this file ?

File: sw/airborne/modules/guidance/gvf/gvf.c [hunk 10]
```diff
@@ -457,3 +476,259 @@ bool gvf_sin_wp_alpha(uint8_t wp, float alpha, float w, float off, float A)
...
+  nav_rover_base.goto_wp.from.y = y1;
+  nav_rover_base.goto_wp.to.x = x2;
+  nav_rover_base.goto_wp.to.y = y2;
+
+  return true;
+}
+
+bool gvf_line_wp1_wp2(uint8_t wp1, uint8_t wp2)
+{
+  float x1 = waypoints[wp1].enu_f.x;
```
Comment (gautierhattenberger) [inline sw/airborne/modules/guidance/gvf/gvf.c:539]: I recommend to use the `waypoint_get_x|y` function instead of accessing the element directly

**Review by gautierhattenberger — CHANGES_REQUESTED** [review]

**Comment by noether** [comment]
Cool, thanks Gautier! 

We will revise the code as suggested.

File: conf/modules/board_matek_f765_car.xml (hunk 2 above)
Comment (jesusBV20) [reply conf/modules/board_matek_f765_car.xml:3]: There are no differences. I had to remove that file

File: sw/airborne/firmwares/rover/main.c (hunk 9 above)
Comment (jesusBV20) [reply sw/airborne/firmwares/rover/main.c:1]: If I don't add this main.c then I get this error when trying to compile with NPS target:

make[1]: *** No rule to make target '/home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main.c', needed by '/home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main.o'.  Stop.
make[1]: *** Waiting for unfinished jobs....
CC /home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main_ap.o
make[1]: Leaving directory '/home/jesus/workspace/repos/paparazzi/sw/airborne'
make: *** [Makefile.ac:213: nps.compile] Error 2

File: sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c (hunk 8 above)
Comment (jesusBV20) [reply sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c]: Cool! I didn't know that

**Comment by noether** [comment]
@gautierhattenberger we will let you know when this PR is ready again. We will reorganize the modes to have manual and auto (we will remove for now the assisted one). We will add a speed controller that we already do not have, and finally, we will see what to do with the telemetry, if we need any extra messages or not.

**Comment by noether** [comment]
@gautierhattenberger  can you check whether we have addressed your suggestions correctly? Maybe we have forgotten something. It has been a long time ago and we have rebased all the (new) commits that were done in parallel from master too.

**Comment by noether** [comment]
There are two new main functionalities in this update:

- We have added a PID (+ feedforward) controller for the control of the ground speed.
- Three modes: 1. Manual, 2. Cruise (it sets as desired velocity the last one measured) but the course is still manual, 3. Auto (course and speed are both autonomous).

@jesusBV20 , let us know whether there are any other major changes.

**Comment by gautierhattenberger** [comment]
There must be something wrong with your merge or rebase on master branch with more than one thousand files changed. Can you check that first ? I can hardly see the changes.

**Comment by jesusBV20** [comment]
I think that I did something wrong when I tried to rebase from our master. We don't know how to undo that, so I have created a clean branch and included every file changed.

**Comment by noether** [comment]
Yep, we really do not how to do it (we are not deep expertise with git xD). 

Because the branch was created a time ago, in order to make sure it will compile with the current master (no subsystems, etc), we pulled (rebase) the current master to the branch. Is there a better way, or less messy, to do this procedure? (I would love to know it because I have a branch that I created 2 years ago xD that I will like to PR soon).

In parallel, @jesusBV20 has managed to create a new branch from the recent master, so there is no problem creating a new PR.

Let us know what would be the most convenient.

**Comment by noether** [comment]
Oki, we were doing some research about rebase, and merge. We cannot revert the changes in this branch. We will create a new PR that you can track easily :P. I will close this PR then.

**Pull request closed without merging**